pkg/ottl/ottlfuncs/func_extract_grok_patterns.go

```diff
@@ -0,0 +1,98 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+package ottlfuncs // import "github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl/ottlfuncs"
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/elastic/go-grok"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/pkg/ottl"
+)
+
+type ExtractGrokPatternsArguments[K any] struct {
+	Target             ottl.StringGetter[K]
+	Pattern            string
+	NamedCapturesOnly  ottl.Optional[bool]
+	PatternDefinitions ottl.Optional[[]ottl.StringGetter[K]]
+}
+
+func NewExtractGrokPatternsFactory[K any]() ottl.Factory[K] {
+	return ottl.NewFactory("ExtractGrokPatterns", &ExtractGrokPatternsArguments[K]{}, createExtractGrokPatternsFunction[K])
+}
+
+func createExtractGrokPatternsFunction[K any](_ ottl.FunctionContext, oArgs ottl.Arguments) (ottl.ExprFunc[K], error) {
+	args, ok := oArgs.(*ExtractGrokPatternsArguments[K])
+
+	if !ok {
+		return nil, fmt.Errorf("ExtractGrokPatternsFactory args must be of type *ExtractGrokPatternsArguments[K]")
+	}
+
+	return extractGrokPatterns(args.Target, args.Pattern, args.NamedCapturesOnly, args.PatternDefinitions)
+}
+
+func extractGrokPatterns[K any](target ottl.StringGetter[K], pattern string, nco ottl.Optional[bool], patternDefinitions ottl.Optional[[]ottl.StringGetter[K]]) (ottl.ExprFunc[K], error) {
+	g := grok.NewComplete()
+	namedCapturesOnly := !nco.IsEmpty() && nco.Get()
+
+	return func(ctx context.Context, tCtx K) (any, error) {
+		if !patternDefinitions.IsEmpty() {
+			getters := patternDefinitions.Get()
+
+			for i, getter := range getters {
+				val, err := getter.Get(ctx, tCtx)
+				if err != nil {
+					return nil, fmt.Errorf("failed to get pattern supplied to ExtractGrokPatterns at index %d: %w", i, err)
+				}
+				// split pattern in format key=val
+				parts := strings.SplitN(val, "=", 2)
+				if len(parts) == 1 {
+					return nil, fmt.Errorf("pattern supplied to ExtractGrokPatterns at index %d has incorrect format, expecting PATTERNNAME=pattern definition", i)
+				}
+
+				g.AddPattern(parts[0], parts[1])
+			}
+		}
+
+		err := g.Compile(pattern, namedCapturesOnly)
```

Comment:
since pattern is static, I believe this can be moved outside the returned function so that the pattern can be validated during startup instead of hot-path execution.

Reply:
yes but then additional patternDefinitions would not be considered as they are optional getters.
we would need to end up with recompile again.
do you think as `g` is outside of func scope, that guarding this with e.g sync.Once would help?

Reply:
Ya I was hoping to not have to recompile. Is it required that `PatternDefinitions` be able to set dynamic patterns or could they be static like `Pattern`?

Reply:
I am guess re-compiling is expensive, so I'd prefer to start with static []string so that we can compile once

Reply:
Adding dynamic string support later wouldn't be a breaking change.